nexus/types/src/deployment/blueprint_diff.rs

```diff
@@ -85,6 +692,58 @@ impl ZoneSortKey for ModifiedZone {
 }
 
 impl ModifiedZone {
+    pub fn from_diff(
+        diff: &BlueprintZoneConfigDiff,
+    ) -> Result<ModifiedZone, BpDiffZoneError> {
+        // Do we have any errors? If so, create a "reason" string.
+        let mut reason = String::new();
+        // These first two checks are only for backwards compatibility. They are
+        // all included in the zone_type comparison below.
+        if diff.zone_type.before.kind() != diff.zone_type.after.kind() {
+            let msg = format!(
+                "mismatched zone kind: before: {}, after: {}\n",
+                diff.zone_type.before.kind().report_str(),
+                diff.zone_type.after.kind().report_str(),
+            );
+            reason.push_str(&msg);
+        }
+        if diff.zone_type.before.underlay_ip()
+            != diff.zone_type.after.underlay_ip()
+        {
+            let msg = format!(
+                "mismatched underlay IP: before: {}, after: {}\n",
+                diff.zone_type.before.underlay_ip(),
+                diff.zone_type.after.underlay_ip()
+            );
+            reason.push_str(&msg);
+        }
+        if diff.zone_type.before != diff.zone_type.after {
+            let msg = format!(
+                "mismatched zone type: after: {:#?}\n",
+                diff.zone_type.after
+            );
+            reason.push_str(&msg);
+        }
+        if reason.is_empty() {
+            Ok(ModifiedZone {
+                prior_disposition: *diff.disposition.before,
+                zone: BlueprintZoneConfig {
+                    disposition: *diff.disposition.after,
+                    id: *diff.id.after,
+                    filesystem_pool: diff.filesystem_pool.after.clone(),
+                    zone_type: diff.zone_type.after.clone(),
+                },
+            })
+        } else {
+            Err(BpDiffZoneError {
+                zone_before_id: *diff.id.before,
+                zone_after_id: *diff.id.after,
+                reason,
+            })
+        }
+    }
+
+    // TODO: Legacy - remove
```

Comment:
Can we remove this?

Reply:
Done in [69fe736](https://github.com/oxidecomputer/omicron/pull/7466/commits/69fe736d8238c3b75d152f73d053ec9b15c6e7da)